numpy/_core/src/multiarray/stringdtype/static_string.c

```diff
@@ -0,0 +1,745 @@
+{
+    PyThread_release_lock(allocator->allocator_lock);
+}
+
+/*NUMPY_API
+ * Release the mutexes locking N allocators.
+ *
+ * If pointers to the same allocator are passed multiple times, only releases
+ * the allocator mutex once.
+ */
+NPY_NO_EXPORT void
+NpyString_release_allocators(size_t n_allocators, npy_string_allocator *allocators[])
+{
+    for (size_t i=0; i<n_allocators; i++) {
+        int matches = 0;
+        for (size_t j=0; j<i; j++) {
+            if (allocators[i] == allocators[j]) {
+                matches = 1;
+                break;
+            }
+        }
+        if (!matches) {
+            NpyString_release_allocator(allocators[i]);
+        }
+    }
+}
+
+
+/*NUMPY_API
+ * Extract the packed contents of *packed_string* into *unpacked_string*.
+ *
+ * The *unpacked_string* is a read-only view onto the *packed_string* data and
+ * should not be used to modify the string data. If *packed_string* is the
+ * null string, sets *unpacked_string.buf* to the NULL pointer. Returns -1 if
+ * unpacking the string fails, returns 1 if *packed_string* is the null
+ * string, and returns 0 otherwise.
+ *
+ * A useful pattern is to define a stack-allocated npy_static_string instance
+ * initialized to {0, NULL} and pass a pointer to the stack-allocated unpacked
+ * string to this function.  This function can be used to simultaneously
+ * unpack a string and determine if it is a null string.
+ */
+NPY_NO_EXPORT int
+NpyString_load(npy_string_allocator *allocator,
+               const npy_packed_static_string *packed_string,
+               npy_static_string *unpacked_string)
+{
+    if (NpyString_isnull(packed_string)) {
+        unpacked_string->size = 0;
+        unpacked_string->buf = NULL;
+        return 1;
+    }
+
+    _npy_static_string_u *string_u = (_npy_static_string_u *)packed_string;
+
+    if (is_short_string(packed_string)) {
+        unsigned char high_byte = string_u->direct_buffer.size_and_flags;
+        unpacked_string->size = high_byte & NPY_SHORT_STRING_SIZE_MASK;
+        unpacked_string->buf = string_u->direct_buffer.buf;
+    }
+
+    else {
+        size_t size = VSTRING_SIZE(string_u);
+        char *buf = NULL;
+        if (size > 0) {
+            npy_string_arena *arena = &allocator->arena;
+            if (arena == NULL) {
+                return -1;
+            }
+            buf = vstring_buffer(arena, string_u);
+            if (buf == NULL) {
+                return -1;
+            }
+        }
+        unpacked_string->size = size;
+        unpacked_string->buf = buf;
+    }
+
+    return 0;
+}
+
+static char *
+heap_or_arena_allocate(npy_string_allocator *allocator,
+                       _npy_static_string_u *to_init_u, size_t size,
+                       int *on_heap)
+{
+    unsigned char *flags = &to_init_u->direct_buffer.size_and_flags;
+    if (*flags & NPY_STRING_SHORT) {
+        // Have to heap allocate since there isn't a preexisting
```

Comment:
You get here if a user mutates an existing short string entry in an array to a string that cannot fit inside a short string. Like so:

```
>>> arr = np.array(["hello", "world"])
>>> arr[0] = "this string is long enough to live on the heap"
```

Because the original string is short, there isn't a prior space in the arena allocation for this string, so we definitely have to malloc a buffer on the heap directly.

Reply:
OK, thanks! I'll have to try to follow the logic for the string replacement a bit better...  (note that I don't think this part is in any way critical, as none of it is exposed).